script/util/rule.js

```diff
@@ -82,7 +82,7 @@ export function rule(filePath) {
       info = JSON.parse(lines[0])
       lines.splice(0, 1)
       /* c8 ignore next 5 */
-    } catch (error) {
+    } catch (/** @type any */ error) {
```

Comment:
This is needed because of a new TypeScript release. It’s not related to the changes.

Reply:
is this truly `any`?
would `unknown` or `Error` work?

Reply:
Catch variables can be either `unknown` or `any`, nothing else. TypeScript switched catch variables from `any` to `unknown` by default in strice mode. ([blog post](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#use-unknown-catch-variables))

It can’t be unknown, because then `error.stack` can’t be accessed on line 87. I also couldn’t cast it to `Error` using jsdoc based types :thinking: 

Alternatively `"useUnknownInCatchVariables": false` could be set in `tsconfig.json` to revert back to the old behaviour, although it makes sense to me to have catch variables defined as `unknown` by default.

Reply:
hmmm :thinking: 
what about

```js
/**
  * @param {unknown} err
  * @return {err is Error}
  */
function isErrorWithStack(err) {
    return typeof err === 'object' && err !== null && 'stack' in err;
}
```

```suggestion
    } catch (error) {
```

then

```js
      'Could not parse example in ' + ruleId + ':\n' + (isErrorWithStack(error) ? error.stack : '')

Reply:
or

```js
/**
  * @param {unknown} err
  * @return {err is Error}
  */
function isError(error) {
    return error instanceof Error;
}

Reply:
TS changed I'd from any to unknown in v4.5.
I've revelry instead casted it on the first line of the catch clause:

Const exception = /* @type {Error} */ (error)

- I like it because it doesn't use any
- I think the cast is fine because it's most probably an actual error object that's thrown

Reply:
Even for actual error stack could be undefined I believe, per TS. So that can be handled with (error.stack || '')

Reply:
None of the following work:

```js
try {
  // …
} catch (/** @type {Error} */ error) {
  const exception = error
  throw new Error(
    'Could not parse example in ' + ruleId + ':\n' + exception.stack
  )
}
```

```js
try {
  // …
} catch (error) {
  /** @type {Error} */
  const exception = error
  throw new Error(
    'Could not parse example in ' + ruleId + ':\n' + exception.stack
  )
}
```

```js
try {
  // …
} catch (error) {
  const exception = /** @type {Error} */ error
  throw new Error(
    'Could not parse example in ' + ruleId + ':\n' + exception.stack
  )
}
```

```js
try {
  // …
} catch (error) {
  const exception = /** @type {Error} */ (error)
  throw new Error(
    'Could not parse example in ' + ruleId + ':\n' + exception.stack
  )
}
```

```js
try {
  // …
} catch (error) {
  const exception = (/** @type {Error} */ error)
  throw new Error(
    'Could not parse example in ' + ruleId + ':\n' + exception.stack
  )
}
```

```js
try {
  // …
} catch (error) {
  throw new Error(
    'Could not parse example in ' + ruleId + ':\n' + (/** @type {Error} */ error).stack
  )
}
```

Basically the TypeScript feature we want is this:

```ts
try {
  // …
} catch (error) {
  throw new Error(
    'Could not parse example in ' + ruleId + ':\n' + (error as Error).stack
  )
}
```

But this isn’t possible.

I think creating a function to check for an error or using an `instanceof` check as @ChristianMurphy proposed would be the best solution. However, I would also like to point out that both occurrences apply to 2 very small catch clauses, which are both in test files. I don’t worry about these 2 values being `any`.

This is a reminder though that TypeScript minor versions may introduce breaking changes in strict mode. For projects not using lock files it may be best to stick with tilde (`~`) versions for TypeScript, not caret (`^`) ranges.

Reply:
4 should work. It makes it into an error. But not all errors have a stack as a string, right? What's the error you get?

Reply:
Must have been a mistake on my side at the time. I tried 4 again, and it works now.